src/components/QueryEditor/VisualQueryEditor/KQLFilter.tsx

```diff
@@ -0,0 +1,112 @@
+import React, { useState } from 'react';
+
+import { SelectableValue } from '@grafana/data';
+import { EditorList } from '@grafana/ui';
+
+import { AdxDataSource } from '../../../datasource';
+import { AdxColumnSchema, KustoQuery } from '../../../types';
+import {
+  QueryEditorArrayExpression,
+  QueryEditorExpression,
+  QueryEditorExpressionType,
+  QueryEditorOperatorExpression,
+} from 'components/LegacyQueryEditor/editor/expressions';
+import { QueryEditorPropertyType } from 'schema/types';
+import { sanitizeOperator } from './utils/utils';
+import FilterItem from './FilterItem';
+
+interface KQLFilterProps {
+  index: number;
+  query: KustoQuery;
+  datasource: AdxDataSource;
+  columns?: AdxColumnSchema[];
+  templateVariableOptions: SelectableValue<string>;
+  onChange: (query: KustoQuery) => void;
+}
+
+const KQLFilter: React.FC<KQLFilterProps> = ({
+  index,
+  query,
+  onChange: onQueryChange,
+  datasource,
+  columns,
+  templateVariableOptions,
+}) => {
+  // Each expression is a group of several OR statements
+  const [filters, setFilters] = useState(
+    (query.expression.where.expressions[index] as QueryEditorArrayExpression)?.expressions
+  );
+
+  const onChange = (newItems: Array<Partial<QueryEditorOperatorExpression>>) => {
+    // As new (empty object) items come in, with need to make sure they have the correct type
+    const cleaned = newItems.map(
+      (v): QueryEditorOperatorExpression => ({
+        type: QueryEditorExpressionType.Operator,
+        property: v.property ?? { type: QueryEditorPropertyType.String, name: '' },
+        operator: v.operator ?? { name: '==', value: '' },
+      })
+    );
+    setFilters(cleaned);
+
+    // Only save valid and complete filters into the query state
+    const validExpressions: QueryEditorOperatorExpression[] = [];
+    for (const operatorExpression of cleaned) {
+      const validated = sanitizeOperator(operatorExpression);
+      if (validated) {
+        validExpressions.push(validated);
+      }
+    }
+
+    const where = { ...query.expression.where };
+    if (newItems.length) {
+      (where.expressions[index] as QueryEditorArrayExpression).expressions = validExpressions;
+    } else {
+      // The expression is empty, remove it
+      const expr = where.expressions;
+      expr.splice(index, 1);
+      where.expressions = expr;
+    }
```

Comment:
explanation: this code only add new expressions to the query if the expression is complete (e.g. it has the three elements, column, operator and value).

In the case that the `change` event is triggered with removing a condition, and there are no more expressions, we remove the entry.

Note: I have found a bug here. If there are multiple groups, removing the first one messes the output :thinking: 

![Peek 2022-08-25 12-54](https://user-images.githubusercontent.com/4025665/186646841-e61c8fe9-48c3-4def-8256-02f1e00b8166.gif)